src/coreclr/jit/gentree.h

```diff
@@ -911,7 +911,7 @@ struct GenTree
     }
 
     // Node is contained, but it isn't contained due to being a containable int.
-    bool isContainedAndNotIntOrIImmed() const
+    bool isContainedCompareChainSegment() const
     {
         return isContained() && !isContainedIntOrIImmed();
```

Comment:
Should also be something like:
```suggestion
        return isContained() && (OperIs(GT_AND) || OperIsCmpCompare());
```
?

I was hoping this could replace the existing `isContainedAndNotIntOrIImmed()`, and the problem would magically be fixed. But maybe that's not the case, I guess `IsValidCompareChain()` needs to work before things have been marked contained.

Reply:
Actually, wouldn't it work?

Reply:
I did go down this route myself at first.

The problem is, in other places we are checking the children of the compare nodes, which won't be `(OperIs(GT_AND) || OperIsCmpCompare())`
For example, see `ContainCheckCompareChain()`

Reply:
or
`if (tree->OperIs(GT_AND) && op2->isContainedCompareChainSegment())`

Reply:
```
    bool isContainedCompareChainSegment(Gentree *parent) const
    {
       if (isContained())
       {
           if (OperIs(GT_AND) || OperIsCmpCompare()) return true;
           if (parent->OperIs(GT_AND) || parent->OperIsCmpCompare()) && !isContainedIntOrIImmed())) return true;
       }
       return false;       
    }
```

That's the nearest I could get

Reply:
> For example, see ContainCheckCompareChain()

Is it handled correctly today? My understanding was that the check there is also supposed to be checking for compare chains, not other contained children (e.g. `AND` with `LSL` child -- should it really be returning true for `ContainCheckCompareChain`?)

Reply:
I.e. I believe that before #75823, the `isContainedAndNotIntOrIImmed` check in the contexts that it was used was equivalent to the `isContained() && (OperIs(GT_AND) || OperIsCmpCompare())`, while after that PR they are not equivalent anymore (as the assertion failure shows).
So to restore the old behavior, can we just switch the implementation of `isContainedAndNotInOrIImmed()` to `isContained() && (OperIs(GT_AND) || OperIsCmpCompare())` (and rename it)? Or am I missing some cases?

Reply:
Ok, updated.

Had to convince myself that isContainedIntOrIImmed() check was no longer required and went through all the use cases of isContainedCompareChainSegment().

isContainedCompareChainSegment() is now being used to check that a parent and its child form part of a compare chain (the parent must be an AND).

(accidentally force pushed, but the old commit remains unchained)

Reply:
Very nice, this looks great.